CWE-754 - Improper Check for Unusual or Exceptional Conditions - Float (#842)
ossf/wg-best-practices-os-developers: docs/Secure-Coding-Guide-for-Python/CWE-703/CWE-754/README.md

# CWE-754: Improper Check for Unusual or Exceptional Conditions - Float

Ensure to have handling for exceptional floating-point values.

> The float class has the capability to interpret various input values as floating-point numbers. Some special cases can interpret input values as

Reviewer: Jut to make it clearer its the name of the class:
```suggestion
The `float` class has the capability to interpret various input values as floating-point numbers. Some special cases can interpret input values as
```

* Positive Infinity
* Negative Infinity
* NaN (Not-a-Number)

These floating-point class values represent numbers that fall outside the typical range and exhibit unique behaviors. `NaN` (Not a Number) lacks a defined order and is not considered equal to any value, including itself. Hence, evaluating an expression such as `NaN == NaN` returns `False`.

## Non-Compliant Code Example

> The `noncompliant01.py` intent is to ensure that adding objects will not exceed a total weight of 100 units. Validation fails as the code is missing to test for exceptional conditions such as `NaN` or `infinity`.

Reviewer: IMO this would sound a bit more natural:
```suggestion
The `noncompliant01.py` intent is to ensure that adding objects will not exceed a total weight of 100 units. Validation fails as the code to test for exceptional conditions, such as `NaN` or `infinity`, is missing.
```

*[noncompliant01.py](noncompliant01.py):*

```py
# SPDX-FileCopyrightText: OpenSSF project contributors
# SPDX-License-Identifier: MIT
"""Non-compliant Code Example"""
 
import sys
 
 
class Package:
    """Class representing a package object"""
 

>     def __init__(self):
>         self.package_weight = float(0.0)
>         self.max_package_weight = 100.0

Reviewer: Since this part doesn't impact the portrayed issue and the `compliant01.py` version is more readable, we might as well use it here:
```suggestion
    def __init__(self):
        self.package_weight: float = 0.0
        self.max_package_weight: float = 100.0
```

 
    def add_to_package(self, object_weight: str):
        """Function for adding an object into the package"""
        value = float(object_weight)
        # This is dead code as value gets type cast to float,
        # hence will never be equal to string "NaN"
        if value == "NaN":
            raise ValueError("'NaN' not a number")
        # This is also dead code as value is type cast to float,
        # unusual inputs like -infinity will not get caught

>         if isinstance(value, float) is not True:

Reviewer: You can use `False` instead to simplify it a little:
```suggestion
        if isinstance(value, float) is False:
```

            raise ValueError("not a number")
        if self.package_weight + value > self.max_package_weight:
            raise ValueError("Addition would exceed maximum package weight.")
 
        self.package_weight += value
 
    def get_package_weight(self):
        """Getter for outputting the package's current weight"""
        return self.package_weight
 
 
#####################
# exploiting above code example
#####################
package = Package()
print(f"\nOriginal package's weight is {package.get_package_weight():.2f} units\n")
for item in [100, "-infinity", sys.float_info.max, "NaN", -100]:
    print(f"package.add_to_package({item})")
    try:
        package.add_to_package(item)
        print(
            f"package.get_package_weight() = {package.get_package_weight():.2f}\n"
        )
 
    except Exception as e:
        print(e)
```

Some important considerations when dealing with floating-point values from `non-complaint01.py`.

> * Setting a value above sys.float_info.max does not increase the held value. In some cases, incrementing `package_weight`  with a high enough value may turn its value into inf.
> * Setting the added value to -infinity and +infinity causes the value of the `package_weight` to be infinite as well.
> * Adding "NaN", which is not a valid value to `package_weight` will always return "nan".

Reviewer: Small formatting cleanup:
```suggestion
* Setting a value above `sys.float_info.max` does not increase the held value. In some cases, incrementing `package_weight`  with a high enough value may turn its value into `inf`.
* Setting the added value to `-infinity` and `infinity` causes the value of the `package_weight` to be infinite as well.
* Adding `NaN`, which is not a valid value to `package_weight` will always return `NaN`.
```

Reviewer 2: I think I got all of them as part of my cosmetic update

        if isnan(value):
            raise ValueError("Input is not a number")
        if not isfinite(value):
            raise ValueError("Input is not a finite number.")
        if value < 0:
            raise ValueError("Weight must be a non-negative number.")
        if self.package_weight + value > self.max_package_weight:
            raise ValueError("Addition would exceed maximum package weight.")
 
        print(f"Adding an object that weighs {value} units to package")
        self.package_weight += value
 
    def get_package_weight(self) -> float:
        """Return the package's current weight."""
        return self.package_weight

#####################
# exploiting above code example
#####################
 
 
package = Package()
print(f"\nOriginal package's weight is {package.get_package_weight():.2f} units\n")
for item in [100, "-infinity", sys.float_info.max, "NaN", -100]:
    print(f"package.add_to_package({item})")
    try:
        package.add_to_package(item)
        print(
            f"package.get_package_weight() = {package.get_package_weight():.2f}\n"
        )
 
    except Exception as e:
        print(e)
```

> This compliant code example will raise a ValueError for inputs that are "-infinity", "infinity", or NaN, with messages "Input is not a finite number" and "Input is not a number" respectively. It should also ensure weights are non-negative, returning "Weight must be a non-negative number" for negative inputs.

Reviewer 2: I think I got all of these as part of my cosmetic updates

# CWE-754: Improper Check for Unusual or Exceptional Conditions - Float

Ensure to have handling for exceptional floating-point values.

The float class has the capability to interpret various input values as floating-point numbers. Some special cases can interpret input values as

* Positive Infinity
* Negative Infinity
* NaN (Not-a-Number)

> TThese floating-point class values represent numbers that fall outside the typical range and exhibit unique behaviors. NaN (Not a Number) lacks a defined order and is not considered equal to any value, including itself. Hence, evaluating an expression such as "NaN == NaN" returns "False."

Reviewer 2: fix typo, use backticks for code

```suggestion
These floating-point class values represent numbers that fall outside the typical range and exhibit unique behaviors. `NaN` (Not a Number) lacks a defined order and is not considered equal to any value, including itself. Hence, evaluating an expression such as `NaN == NaN` returns `False`.
```

## Non-Compliant Code Example

> In the `noncompliant01.py` example, the code allows a user to add objects of a certain weight to a package—as long as the total weight remains below 100 units, the object is added. However, the checks in the code are inadequate for filtering out exceptional conditions. For example, converting the input to a float can mask invalid inputs (such as non-numeric strings), allowing erroneous or unintended values (e.g., NaN or infinity) to bypass validation.

Reviewer 2: remove fluff?

```suggestion
The `noncompliant01.py` intent is to ensure that adding objects will not exceed a total weight of 100 units. Validation fails as the code is missign to test for exceptional conditions such as `NaN` or `infinity`.
```

            raise ValueError("not a number")
        if self.package_weight + value > self.max_package_weight:
            raise ValueError("Addition would exceed maximum package weight.")
 
        self.package_weight += value
 
    def get_package_weight(self):
        """Getter for outputting the package's current weight"""
        return self.package_weight
 
 
#####################
# exploiting above code example
#####################
package = Package()
print(f"\nOriginal package's weight is {package.get_package_weight():.2f} units\n")
for item in [100, "-infinity", sys.float_info.max, "NaN", -100]:
    print(f"package.add_to_package({item})")
    try:
        package.add_to_package(item)
        print(
            f"package.get_package_weight() = {package.get_package_weight():.2f}\n"
        )
 
    except Exception as e:
        print(e)
```

* Setting a value above sys.float_info.max does not increase the held value. In some cases, incrementing package_weight  with a high enough value may turn its value into inf
* Setting the added value to -infinity  and +infinity causes the value of the package_weight  to be infinite as well.

> * Setting the package_weight to  NaN, which is not a valid value.

Reviewer 2: This section should have 
* a sentence before the bullet points
* all code snippets in backticks
* matching variable name

I also don't understand the last bullet.

Author: I added a sentence above the bullet points, I added backticks and I changed the last bullet point, not sure what you meant by "matching variable name"

**Example `noncompliant01.py` output:**

```bash
Original package's weight is 0.00 units

package.add_to_package(100)
package.get_package_weight() = 100.00

package.add_to_package(-infinity)
package.get_package_weight() = -inf

package.add_to_package(1.7976931348623157e+308)
package.get_package_weight() = -inf

package.add_to_package(NaN)
package.get_package_weight() = nan

package.add_to_package(-100)
package.get_package_weight() = nan
```

## Compliant Solution

> Exceptional values and out-of-range values are handled in `compliant01.py`. Due to the nature of the code example, negative values are also checked for.
> The `isfinite` function from the `math` library is useful for checking for `NaN`, `infinity` and `-infinity` values. It checks if a value is neither infinite nor a NaN.

Reviewer 2: suggestion on 2nd sentence, missing some ticks at end.

```suggestion
Exceptional values and out-of-range values are handled in `compliant01.py`. Some negative values are also checked for due to the nature of the code example.
The `isfinite` function from the `math` library is useful for checking for `NaN`, `infinity` and `-infinity` values. `math.isfinite` checks if a value is neither `infinite` nor a `NaN`.
```

On the last two sentences. If `isfinite` is also checking for `isnan` why have both checks in the code?

        if isnan(value):
            raise ValueError("Input is not a number")
        if not isfinite(value):
            raise ValueError("Input is not a finite number.")
        if value < 0:
            raise ValueError("Weight must be a non-negative number.")
        if self.package_weight + value > self.max_package_weight:
            raise ValueError("Addition would exceed maximum package weight.")
 
        print(f"Adding an object that weighs {value} units to package")
        self.package_weight += value
 
    def get_package_weight(self) -> float:
        """Return the package's current weight."""
        return self.package_weight

#####################
# exploiting above code example
#####################
 
 
package = Package()
print(f"\nOriginal package's weight is {package.get_package_weight():.2f} units\n")
for item in [100, "-infinity", sys.float_info.max, "NaN", -100]:
    print(f"package.add_to_package({item})")
    try:
        package.add_to_package(item)
        print(
            f"package.get_package_weight() = {package.get_package_weight():.2f}\n"
        )
 
    except Exception as e:
        print(e)
```

> The compliant code example successfully ensures that any object added to the package is a valid and anticipated float value, and that the code cannot be exploited by inputting values such as NaN, infinite or -infinite.

Reviewer 2: no use to kind of repeat what we say at the very start.
Here we should state what the 'wanted' behaviour is, specifically : compliant code throws and exception .... 

I would also suggest to hadd TODO's to the code to make clearer that the code is an incomplete example. 
```python
# TODO: input sanitation.
# TODO: proper exception handling
```

Author: Updated the sentence and added the "TODOs" now

**Example `compliant01.py` output:**

```bash
Original package's weight is 0.00 units
 
package.add_to_package(100)
Adding an object that weighs 100.0 units to package
package.get_package_weight() = 100.00
 
package.add_to_package(-infinity)
Input is not a finite number.
package.add_to_package(1.7976931348623157e+308)
Addition would exceed maximum package weight.
package.add_to_package(NaN)
Input is not a number
package.add_to_package(-100)
Weight must be a non-negative number.
```

## Automated Detection

|||||
|:---|:---|:---|:---|
|Tool|Version|Checker|Description|
|bandit|1.7.4|no detection||

## Related Guidelines

|||
|:---|:---|
|[SEI CERT C Coding Standard](https://wiki.sei.cmu.edu/confluence/display/c/SEI+CERT+C+Coding+Standard)|[FLP04-C. Check floating-point inputs for exceptional values](https://wiki.sei.cmu.edu/confluence/display/c/FLP04-C.+Check+floating-point+inputs+for+exceptional+values)|

> |[SEI CERT C++ Coding Standard](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=88046682)|[VOID FLP04-CPP. Check floating point inputs for exceptional values](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=88046805)|

Reviewer 2: no longer public reachable by the looks, would remove it
```suggestion
```

|[SEI CERT Oracle Coding Standard for Java](https://wiki.sei.cmu.edu/confluence/display/java/SEI+CERT+Oracle+Coding+Standard+for+Java?src=sidebar)|[NUM08-J. Check floating-point inputs for exceptional values](https://wiki.sei.cmu.edu/confluence/display/java/NUM08-J.+Check+floating-point+inputs+for+exceptional+values)|
|[CWE MITRE Pillar](http://cwe.mitre.org/)|[CWE-703: Improper Check or Handling of Exceptional Conditions](https://cwe.mitre.org/data/definitions/703.html)|
|[MITRE CWE Base](https://cwe.mitre.org/)|[CWE-754: Improper Check for Unusual or Exceptional Conditions](https://cwe.mitre.org/data/definitions/754.html)|

## Biblography

|||
|:---|:---|

> |[NUM07-P3 Do not attempt comparisons with NaN](https://eteamspace.internal.ericsson.com/display/DEVEN/CWE-230%3A+Improper+Handling+of+Missing+Values)| NUM07-P3 Do not attempt comparisons with NaN.|

Reviewer 2: We don't have any quotations/bilbo content, does belong into related guidelines?

Author: I removed the bibliography as the link was to an internal page anyways (oops)!
